firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteIndexManager.java

```diff
@@ -87,6 +105,237 @@ public void addFieldIndex(FieldIndex index) {
         true);
   }
 
+  @Override
+  public void addIndexEntries(Document document) {
+    String collectionGroup = document.getKey().getPath().popLast().canonicalString();
+    db.query(
+            "SELECT index_id, index_proto FROM index_configuration WHERE collection_group = ? AND active = 1")
+        .binding(collectionGroup)
+        .forEach(
+            row -> {
+              try {
+                int indexId = row.getInt(0);
+                FieldIndex fieldIndex =
+                    serializer.decodeFieldIndex(
+                        collectionGroup, row.getInt(0), Index.parseFrom(row.getBlob(1)));
+
+                List<Value> values = extractFieldValue(document, fieldIndex);
+                if (values == null) return;
+
+                List<byte[]> encodeValues = encodeDocumentValues(fieldIndex, values);
+                for (byte[] encoded : encodeValues) {
+                  // TODO(indexing): Handle different values for different users
+                  db.execute(
+                      "INSERT OR IGNORE INTO index_entries ("
+                          + "index_id, "
+                          + "index_value, "
+                          + "document_name) VALUES(?, ?, ?)",
+                      indexId,
+                      encoded,
+                      document.getKey().getPath().canonicalString());
+                }
+              } catch (InvalidProtocolBufferException e) {
+                throw fail("Invalid index: " + e);
+              }
+            });
+  }
+
+  /**
+   * Returns the list of values for all fields indexed by the provided field index. Returns {@code
+   * null} if one or more of the fields are not set.
+   */
+  @Nullable
+  private List<Value> extractFieldValue(Document document, FieldIndex fieldIndex) {
+    List<Value> values = new ArrayList<>();
+    for (FieldIndex.Segment segment : fieldIndex) {
+      Value field = document.getField(segment.getFieldPath());
+      if (field == null) {
+        return null;
+      }
+      values.add(field);
+    }
+    return values;
+  }
+
+  @Override
+  @Nullable
+  public Set<DocumentKey> getDocumentsMatchingTarget(Target target) {
```

Comment:
Can we add some logging here? It might come in handy troubleshooting performance issues going forward.

Reply:
Added logging.